src/hammer-vlsi/hammer_vlsi/hammer_tool.py

```diff
@@ -1064,6 +1064,206 @@ def get_bumps(self) -> Optional[BumpsDefinition]:
             pitch=Decimal(str(self.get_setting("vlsi.inputs.bumps.pitch"))),
             cell=self.get_setting("vlsi.inputs.bumps.cell"), assignments=assignments)
 
+    def generate_visualization(self) -> None:
+        """
+        Generate an SVG visualizing the chip.
+        Depending on mode, it will display floorplan (placement constraints for the current hierarchy)
+        and/or bumps (overlaid on top of floorplan) and pad designators (if this tool is a PCBDeliverableTool).
+        Call this from any custom hook (not used by any default flow).
+        Note: visualizations generated within a PCBDeliverableTool are mirrored about the y-axis (assumption: flip-chip on PCB)
+        """
+        viz_mode = self.get_setting("vlsi.inputs.visualization.mode") #type: str
+        viz_file = self.get_setting("vlsi.inputs.visualization.svg_file") #type: str
+        shorten_path_depth = self.get_setting("vlsi.inputs.visualization.shorten_path_depth") #type: int
+
+        # Checks
+        if not viz_mode in ["all", "floorplan", "bumps", "footprint"]:
+            self.logger.error("Invalid type for 'vlsi.inputs.visualization.type'! No visualization generated.")
+            return
+
+        fp_consts = self.get_placement_constraints() #type: List[PlacementConstraint]
+        # TODO: make pcb action recognize toplevel constraint if par action had hierarchical constraints
+        try:
+            # Get toplevel constraint
+            top = next(filter(lambda x: x.type == PlacementConstraintType.TopLevel, fp_consts)) #type: PlacementConstraint
```

Comment:
I thought we added a helper method for this. If not we certainly should.

Reply:
There is (`get_toplevel` in the hierarchical helper of `HammerDriver`), but I'm pretty sure it's out of scope from the perspective of `HammerTool` because that is used before hierarchical constraints are processed.

Reply:
Hmm, seems like a flaw in that helper method, yet another issue 😢